File: src/JetpacFR.Web/webroot/JetpacFR.Core/CtrlMapModel.js

import { Record, Union } from "../fable_modules/fable-library-js.5.17.0/Types.js";
import { tuple_type, array_type, option_type, string_type, record_type, bool_type, int32_type, union_type } from "../fable_modules/fable-library-js.5.17.0/Reflection.js";
import { min, max } from "../fable_modules/fable-library-js.5.17.0/Double.js";
import { reverse, cons, empty, map as map_2, filter, choose, toArray as toArray_1, tryFind } from "../fable_modules/fable-library-js.5.17.0/List.js";
import { find, contains, item, map as map_1, setItem, fill, max as max_1 } from "../fable_modules/fable-library-js.5.17.0/Array.js";
import { safeHash, equals, disposeSafe, getEnumerator, comparePrimitives } from "../fable_modules/fable-library-js.5.17.0/Util.js";
import { CommentKind } from "../ControlTypesWeb.js";
import { map, delay, toArray } from "../fable_modules/fable-library-js.5.17.0/Seq.js";
import { rangeDouble } from "../fable_modules/fable-library-js.5.17.0/Range.js";
import { Array_distinct } from "../fable_modules/fable-library-js.5.17.0/Seq2.js";
import { disasmMemory } from "./Disasm.js";
import { printf, toText, join } from "../fable_modules/fable-library-js.5.17.0/String.js";

/**
 * Semantic kind of one pixel row (Mixed = several kinds sampled).
 */
export class RowKind extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["UnmappedR", "CodeR", "DataR", "GapR", "MixedR"];
    }
    static UnmappedR = new RowKind(0, []);
    static CodeR = new RowKind(1, []);
    static DataR = new RowKind(2, []);
    static GapR = new RowKind(3, []);
    static MixedR = new RowKind(4, []);
}

export function RowKind_$reflection() {
    return union_type("JetpacFR.Core.CtrlMapModel.RowKind", [], RowKind, () => [[], [], [], [], []]);
}

/**
 * One pixel row of the strip.
 */
export class MapRow extends Record {
    constructor(Kind, Heat, SelfMod, Comment$) {
        super();
        this.Kind = Kind;
        this.Heat = (Heat | 0);
        this.SelfMod = SelfMod;
        this.Comment = Comment$;
    }
}

export function MapRow_$reflection() {
    return record_type("JetpacFR.Core.CtrlMapModel.MapRow", [], MapRow, () => [["Kind", RowKind_$reflection()], ["Heat", int32_type], ["SelfMod", bool_type], ["Comment", bool_type]]);
}

export class LabelInfo extends Record {
    constructor(Addr, Text$) {
        super();
        this.Addr = (Addr | 0);
        this.Text = Text$;
    }
}

export function LabelInfo_$reflection() {
    return record_type("JetpacFR.Core.CtrlMapModel.LabelInfo", [], LabelInfo, () => [["Addr", int32_type], ["Text", string_type]]);
}

/**
 * Instruction-level readout, present only at high zoom.
 */
export class InsnSnippet extends Record {
    constructor(Addr, Bytes, Text$, Comment$) {
        super();
        this.Addr = (Addr | 0);
        this.Bytes = Bytes;
        this.Text = Text$;
        this.Comment = Comment$;
    }
}

export function InsnSnippet_$reflection() {
    return record_type("JetpacFR.Core.CtrlMapModel.InsnSnippet", [], InsnSnippet, () => [["Addr", int32_type], ["Bytes", string_type], ["Text", string_type], ["Comment", option_type(string_type)]]);
}

export class MapRender extends Record {
    constructor(ViewStart, ViewEnd, Rows, Labels, Ranges, Snippets) {
        super();
        this.ViewStart = (ViewStart | 0);
        this.ViewEnd = (ViewEnd | 0);
        this.Rows = Rows;
        this.Labels = Labels;
        this.Ranges = Ranges;
        this.Snippets = Snippets;
    }
}

export function MapRender_$reflection() {
    return record_type("JetpacFR.Core.CtrlMapModel.MapRender", [], MapRender, () => [["ViewStart", int32_type], ["ViewEnd", int32_type], ["Rows", array_type(MapRow_$reflection())], ["Labels", array_type(LabelInfo_$reflection())], ["Ranges", array_type(tuple_type(int32_type, int32_type))], ["Snippets", array_type(InsnSnippet_$reflection())]]);
}

/**
 * Address -> fractional row index.
 */
export function addrY(viewStart, viewEnd, rowCount, addr) {
    return (rowCount * (addr - viewStart)) / max(1, viewEnd - viewStart);
}

/**
 * Top address of pixel row y.
 */
export function rowLo(viewStart, viewEnd, rowCount, y) {
    return (viewStart + ~~(((viewEnd - viewStart) * y) / rowCount)) | 0;
}

function classify(blocks, addr) {
    const matchValue = tryFind((b) => {
        if (b.Start <= addr) {
            return addr < b.EndExcl;
        }
        else {
            return false;
        }
    }, blocks);
    if (matchValue != null) {
        const matchValue_1 = matchValue.Kind;
        switch (matchValue_1.tag) {
            case 1:
                return RowKind.DataR;
            case 2:
                return RowKind.GapR;
            default:
                return RowKind.CodeR;
        }
    }
    else {
        return RowKind.UnmappedR;
    }
}

/**
 * Build the render for the current viewport.
 * 
 * counts/selfMod follow Trace.PerPcCount/SelfModified conventions
 * (indexed by address); pass empty arrays when no trace exists.
 */
export function render(blocks, comments, instrStarts, mem, counts, selfMod, viewStart, viewEnd, rowCount) {
    let memory_1, starts_1, snips, a_4, guard;
    const viewStart_1 = max(0, min(65535, viewStart)) | 0;
    const viewEnd_1 = max(viewStart_1 + 1, min(65536, viewEnd)) | 0;
    const span = (viewEnd_1 - viewStart_1) | 0;
    const ppb = rowCount / span;
    const maxC = ((counts.length === 0) ? 1 : max(1, max_1(counts, {
        Compare: (x, y) => (comparePrimitives(x, y) | 0),
    }))) | 0;
    const logMax = Math.log10(maxC);
    const covered = fill(new Array(65536), 0, 65536, false);
    const enumerator = getEnumerator(comments);
    try {
        while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
            const m = enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]();
            if (!equals(m.Kind, CommentKind.Frames)) {
                const hi = (equals(m.Kind, CommentKind.Line) ? (m.Addr + 1) : max(m.Addr, m.EndExcl)) | 0;
                const lo = min(65535, m.Addr) | 0;
                const hi_1 = min(65536, hi) | 0;
                for (let a_1 = lo; a_1 <= (hi_1 - 1); a_1++) {
                    setItem(covered, a_1, true);
                }
            }
        }
    }
    finally {
        disposeSafe(enumerator);
    }
    const samples = toArray(delay(() => map((i) => (~~((i * span) / 8) | 0), rangeDouble(0, 1, 7))));
    const rows = fill(new Array(rowCount), 0, rowCount, null);
    for (let y_1 = 0; y_1 <= (rowCount - 1); y_1++) {
        const lo_1 = min(rowLo(viewStart_1, viewEnd_1, rowCount, y_1), viewEnd_1 - 1) | 0;
        const hi_2 = max(lo_1 + 1, rowLo(viewStart_1, viewEnd_1, rowCount, y_1 + 1)) | 0;
        const distinct = Array_distinct(map_1((off) => classify(blocks, min(65535, lo_1 + off)), samples), {
            Equals: equals,
            GetHashCode: (x_1) => (safeHash(x_1) | 0),
        });
        const kind = (distinct.length === 1) ? item(0, distinct) : (((distinct.length === 2) && contains(RowKind.UnmappedR, distinct, {
            Equals: equals,
            GetHashCode: (x_2) => (safeHash(x_2) | 0),
        })) ? find((y_4) => !equals(RowKind.UnmappedR, y_4), distinct) : RowKind.MixedR);
        const heat = max_1(map_1((off_1) => {
            const a = (lo_1 + off_1) | 0;
            const c = ((a < counts.length) ? item(a, counts) : 0) | 0;
            if (c <= 0) {
                return 0;
            }
            else {
                return min(9, ~~(9 * (Math.log10(c) / logMax))) | 0;
            }
        }, samples, Int32Array), {
            Compare: (x_4, y_5) => (comparePrimitives(x_4, y_5) | 0),
        }) | 0;
        const selfModHit = samples.some((off_2) => {
            const a_2 = (lo_1 + off_2) | 0;
            if (a_2 < selfMod.length) {
                return item(a_2, selfMod);
            }
            else {
                return false;
            }
        });
        let hasComment;
        const a0 = min(65535, lo_1) | 0;
        const a1 = min(65535, hi_2 - 1) | 0;
        hasComment = (item(a0, covered) ? true : item(a1, covered));
        setItem(rows, y_1, new MapRow(kind, heat, selfModHit, hasComment));
    }
    return new MapRender(viewStart_1, viewEnd_1, rows, toArray_1(choose((b) => {
        const y0 = addrY(viewStart_1, viewEnd_1, rowCount, b.Start);
        if ((((addrY(viewStart_1, viewEnd_1, rowCount, b.EndExcl) - y0) >= 10) && (b.Start < viewEnd_1)) && (b.EndExcl > viewStart_1)) {
            return new LabelInfo(b.Start, b.Name);
        }
        else {
            return undefined;
        }
    }, blocks)), toArray_1(filter((tupledArg) => (tupledArg[1] > tupledArg[0]), map_2((m_2) => [max(viewStart_1, m_2.Addr), min(viewEnd_1, m_2.EndExcl)], filter((m_1) => {
        if (equals(m_1.Kind, CommentKind.Range$)) {
            return m_1.EndExcl > m_1.Addr;
        }
        else {
            return false;
        }
    }, comments)))), (instrStarts != null) ? ((mem != null) ? ((ppb >= 8) ? ((memory_1 = mem, (starts_1 = instrStarts, (snips = empty(), (a_4 = viewStart_1, (guard = 0, ((() => {
        while ((a_4 < viewEnd_1) && (guard < 256)) {
            guard = ((guard + 1) | 0);
            while (((a_4 < viewEnd_1) && (a_4 < 65536)) && !item(a_4, starts_1)) {
                a_4 = ((a_4 + 1) | 0);
            }
            if ((a_4 < viewEnd_1) && (a_4 < 65536)) {
                const insn = disasmMemory(memory_1, a_4);
                const len = max(1, insn.Length) | 0;
                const hex = join(" ", delay(() => map((i_1) => {
                    const arg = item((a_4 + i_1) & 65535, memory_1);
                    return toText(printf("%02X"))(arg);
                }, rangeDouble(0, 1, len - 1))));
                let cmt;
                const option_1 = tryFind((m_3) => {
                    if (equals(m_3.Kind, CommentKind.Line)) {
                        return m_3.Addr === a_4;
                    }
                    else {
                        return false;
                    }
                }, comments);
                cmt = ((option_1 != null) ? option_1.Text : undefined);
                snips = cons(new InsnSnippet(a_4, hex, insn.Text, cmt), snips);
                a_4 = ((a_4 + len) | 0);
            }
        }
    })(), toArray_1(reverse(snips))))))))) : []) : []) : []);
}

File: src/JetpacFR.Core/CtrlMapModel.fs
namespace JetpacFR.Core

open System

/// Pure layout model behind the desktop ControlMap panel: turns the
/// control file's structure plus optional trace/memory overlays into a
/// per-pixel-row description the renderer just paints. No WPF types here,
/// so the same geometry can drive a canvas renderer later.
///
/// The vertical axis is the address space: row 0 is viewStart, row
/// rowCount-1 the top of viewEnd's pixel. Address <-> row mapping is
/// linear through `addrY` / `rowLo` / `rowHi`.
module CtrlMapModel =

    /// Semantic kind of one pixel row (Mixed = several kinds sampled).
    type RowKind =
        | UnmappedR
        | CodeR
        | DataR
        | GapR
        | MixedR

    /// One pixel row of the strip.
    type MapRow =
        {
            Kind: RowKind
            /// 0-9 log-scaled execution heat (0 = never executed).
            Heat: int
            SelfMod: bool
            /// Any Line comment hit or Name/Range cover inside the row.
            Comment: bool
        }

    type LabelInfo = { Addr: int; Text: string }

    /// Instruction-level readout, present only at high zoom.
    type InsnSnippet =
        { Addr: int
          Bytes: string
          Text: string
          Comment: string option }

    type MapRender =
        {
            ViewStart: int
            ViewEnd: int
            Rows: MapRow[]
            /// Visible block-name labels (band tall enough to hold one line).
            Labels: LabelInfo[]
            /// Visible range-comment extents as address pairs.
            Ranges: (int * int)[]
            Snippets: InsnSnippet[]
        }

    /// Address -> fractional row index.
    let addrY (viewStart: int) (viewEnd: int) (rowCount: int) (addr: int) : float =
        float rowCount * float (addr - viewStart) / float (max 1 (viewEnd - viewStart))

    /// Top address of pixel row y.
    let rowLo (viewStart: int) (viewEnd: int) (rowCount: int) (y: int) : int =
        viewStart + (viewEnd - viewStart) * y / rowCount

    let private classify (blocks: Block list) (addr: int) : RowKind =
        match blocks |> List.tryFind (fun b -> b.Start <= addr && addr < b.EndExcl) with
        | None -> UnmappedR
        | Some b ->
            match b.Kind with
            | Code -> CodeR
            | Data -> DataR
            | Gap -> GapR

    /// Build the render for the current viewport.
    ///
    /// counts/selfMod follow Trace.PerPcCount/SelfModified conventions
    /// (indexed by address); pass empty arrays when no trace exists.
    let render
        (blocks: Block list)
        (comments: ControlComment list)
        (instrStarts: bool[] option)
        (mem: byte[] option)
        (counts: int[])
        (selfMod: bool[])
        (viewStart: int)
        (viewEnd: int)
        (rowCount: int)
        : MapRender =

        let viewStart = max 0 (min 0xFFFF viewStart)
        let viewEnd = max (viewStart + 1) (min 0x10000 viewEnd)
        let span = viewEnd - viewStart
        let ppb = float rowCount / float span

        // Heat scale: same log mapping as refreshHeatmap so both views agree.
        let maxC = if counts.Length = 0 then 1 else max 1 (Array.max counts)
        let logMax = log10 (float maxC)

        let level (a: int) =
            let c = if a < counts.Length then counts[a] else 0

            if c <= 0 then
                0
            else
                min 9 (int (9.0 * (log10 (float c) / logMax)))

        // Comment coverage bitmap: Line hits plus everything a Name/Range spans.
        // Frames comments cover trace-state ranges, not addresses - skipped.
        let covered = Array.create 0x10000 false

        for m in comments do
            if m.Kind <> CommentKind.Frames then
                let hi =
                    if m.Kind = CommentKind.Line then
                        m.Addr + 1
                    else
                        max m.Addr m.EndExcl

                let lo = min 0xFFFF m.Addr
                let hi = min 0x10000 hi

                for a in lo .. hi - 1 do
                    covered[a] <- true

        // Row classification by sampling; 8 probes resolve mixed content well
        // enough at every zoom the renderer allows.
        let samples = [| for i in 0..7 -> float i * float span / 8.0 |> int |]
        let rows = Array.zeroCreate<MapRow> rowCount

        for y in 0 .. rowCount - 1 do
            // When zoomed in past 1 byte/pixel several rows share one address;
            // clamp to a non-empty probe range instead of inheriting a neighbor.
            let lo = min (rowLo viewStart viewEnd rowCount y) (viewEnd - 1)
            let hi = max (lo + 1) (rowLo viewStart viewEnd rowCount (y + 1))

            let kinds =
                samples |> Array.map (fun off -> classify blocks (min 0xFFFF (lo + off)))

            let distinct = kinds |> Array.distinct

            let kind =
                if distinct.Length = 1 then
                    distinct[0]
                elif distinct.Length = 2 && distinct |> Array.contains UnmappedR then
                    distinct |> Array.find ((<>) UnmappedR)
                else
                    MixedR

            let heat = samples |> Array.map (fun off -> level (lo + off)) |> Array.max

            let selfModHit =
                samples
                |> Array.exists (fun off ->
                    let a = lo + off
                    a < selfMod.Length && selfMod[a])

            let hasComment =
                let a0 = min 0xFFFF lo
                let a1 = min 0xFFFF (hi - 1)
                covered[a0] || covered[a1]

            rows[y] <-
                { Kind = kind
                  Heat = heat
                  SelfMod = selfModHit
                  Comment = hasComment }

        let labels =
            blocks
            |> List.choose (fun b ->
                let y0 = addrY viewStart viewEnd rowCount b.Start
                let y1 = addrY viewStart viewEnd rowCount b.EndExcl

                if y1 - y0 >= 10.0 && b.Start < viewEnd && b.EndExcl > viewStart then
                    Some { Addr = b.Start; Text = b.Name }
                else
                    None)
            |> Array.ofList

        // Range-comment extents clipped to the viewport.
        let ranges =
            comments
            |> List.filter (fun m -> m.Kind = CommentKind.Range && m.EndExcl > m.Addr)
            |> List.map (fun m -> (max viewStart m.Addr, min viewEnd m.EndExcl))
            |> List.filter (fun (a, b) -> b > a)
            |> Array.ofList

        // Instruction-level readouts past the readability threshold.
        let snippets =
            match instrStarts, mem with
            | Some starts, Some memory when ppb >= 8.0 ->
                let mutable snips = []
                let mutable a = viewStart
                let mutable guard = 0

                while a < viewEnd && guard < 256 do
                    guard <- guard + 1

                    while a < viewEnd && a < 0x10000 && not starts[a] do
                        a <- a + 1

                    if a < viewEnd && a < 0x10000 then
                        let insn = Disasm.disasmMemory memory a
                        let len = max 1 insn.Length

                        let hex =
                            seq { for i in 0 .. len - 1 -> sprintf "%02X" memory[(a + i) &&& 0xFFFF] }
                            |> String.concat " "

                        let cmt =
                            comments
                            |> List.tryFind (fun m -> m.Kind = CommentKind.Line && m.Addr = a)
                            |> Option.map (fun m -> m.Text)

                        snips <-
                            { Addr = a
                              Bytes = hex
                              Text = insn.Text
                              Comment = cmt }
                            :: snips

                        a <- a + len

                snips |> List.rev |> Array.ofList
            | _ -> [||]

        { ViewStart = viewStart
          ViewEnd = viewEnd
          Rows = rows
          Labels = labels
          Ranges = ranges
          Snippets = snippets }
